Adds the TX-54 Grenade launcher

Lumipharon opened the pull request:
<!-- ***STOP!***  Read this: If this is not a PR ready for review and merge or WIP, open it as a draft PR, using the arrow next to 'Create Pull Request'>

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<del>Adds the TX-55 Objective Individual Combat Weapon, featuring a bottom mounted small capacity rifle and a top mounted, 8 round magazine fed grenade launcher chambered for 20mm smart grenades.</del>

TX-55 is kill, but will appear elsewhere on occasion via another PR. Leaving in the code/sprites here as that still makes sense though.

Adds the TX-54, essentially the grenade launcher part of the TX-55 but as a stand alone weapon. Slightly less cumbersome due to the lower weight.

Shout out to everyone that lent a hand with this. I am not a spriter, so the sprites I had to do may be... questionable. I am also barely a coder, but everything does work at least.

![image](https://user-images.githubusercontent.com/7869430/156973941-f44aa05f-9961-4a96-b957-0a8dd5b0f61d.png)

TX-55 Loaded and unloaded:
![image](https://user-images.githubusercontent.com/7869430/156974775-bb55423f-5313-4e8a-88ce-46b4e18ecfe1.png)

TX-54
![image](https://user-images.githubusercontent.com/7869430/158048118-590569dc-8dcb-48f8-917b-211a181d56cc.png)

Airburst grenades:
https://user-images.githubusercontent.com/7869430/156975012-50674172-8f43-4bf1-ae61-d0fd7c00ba18.mp4

Incendiary grenades:
https://user-images.githubusercontent.com/7869430/156975055-94cf6b01-0371-448b-a6a6-7a95c803cb52.mp4

Video editing is for nerds.


**What it does**
This gun is based on the XM29 and other similar weapons systems, and reflects the compromises it comes with.

The grenade launcher is the main draw, currently featuring airbursting 20mm grenades of either the regular shrapnel variety, or incendiary. Regular HE are also available through req, but these might not stay after testing.

The grenades detonate on contact with a target, at max range or when they hit the tile you targetted. They then explode into a directional, short range shower of smaller piercing projectiles causing devestating FUN and friendly fire to anything in that general direction.

In exchange however, the rifle is fairly underwhelming, comparable to a T12 in damage output, but has a smaller 36 round magazine, less accuracy and due to the significant extra weight, is much less wieldy.

The damage output of the grenades (should) be fairly decent, especially against tight groups of benos, however expect to burn through ammo like no tomorrow, and then have to scream at the RO for more. You'll also likely have storage issues due to having to juggle two sets of small capacity magazines.

24/03/22
Added the T49 scope to the TX-54 and TX-55. Not compatible with any other weapon currently.

- The scope provides low light visibility when scoped exactly like optical imaging glasses, and enabled aim mode on the gun.
- Also acts as a scope :^).
- Significantly, I've coded in the capability for aim fire for attachment weapons where **both** the main gun and attachment gun have aim mode. This means the TX-55 will be able to aim mode with both firemodes.

**Availability**
Two TX-54's, plus 10 airburst and 4 incendiary magazines are available round start in the specialised tab.

The TX-54 is also available at req for 30 points, with grenade mags are available at 4 points for airburst mags, 6 points for incendiary, and 10 for the HE.

**Random stuff**
This PR also:

-**Removes impact grenades round start**
- Fixes a bug with grenade projectiles always facing north,
- Adds a new ammo behavior to allow for projectiles to pass through mobs but not other objects like windows/shutters/etc (the grenade bonus projectiles use this),
- Adds a new ammo proc based off the existing bonus projectile proc, but divorces it from giving the bonus projectiles extra ammo based off gun stats, and randomises scatter rather than having a fixed fan formation and also removed the 45 degree cap, if anyone wants to make omni direction mines/rockets/etc (The bonus projectiles for the grenades have a randomised scatter to give them some unpredictability, but if it turns out to be shit/not balanced I can change it back to a fixed fan formation.
- Allows aim_mode to apply to attached weapons, where the attached weapon inherently has access to aim mode (so no change to existing attachments)


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Big gun go boom boom, tell marines skill issue when you shoot them in the back and fill them with shrap
Adds a gun with a somewhat new feature. May or may not be balanced.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Adds the TX-54, a magazine fed grenade launcher utilising 20mm airbursting grenades
remove: Impact grenades are no longer available roundstart
fix: grenades fired from weapons no longer always point north mid flight
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

novaepee commented:
TX family grows.

novaepee commented on code/modules/projectiles/magazines/rifles.dm:
```diff
@@ -343,3 +343,38 @@
+	max_rounds = 5
+	icon_state_mini = "mag_sniper"
+
+/obj/item/ammo_magazine/rifle/tx29launcher/he
```
![image](https://user-images.githubusercontent.com/6610922/156978856-e72156a3-e9b4-49a0-bffd-4700bfa66327.png)

Only two references of /obj/item/ammo_magazine/rifle/tx29launcher/he, which means it's not in /obj/item/storage/box/tx29_system nor in req.

novaepee commented on code/game/objects/items/storage/marine_boxes.dm:
```diff
@@ -171,8 +171,26 @@
+	foldable = null
 
-
+/obj/item/storage/box/tx29_system/Initialize(mapload, ...)
```
No /obj/item/ammo_magazine/rifle/tx29launcher/he in crate.

novaepee commented on code/modules/reqs/supplypacks.dm:
```diff
@@ -207,6 +207,24 @@ WEAPONS
+	contains = list(/obj/item/ammo_magazine/rifle/tx29launcher)
+	cost = 5
+	available_against_xeno_only = TRUE
+
```
Impossible to purchase /obj/item/ammo_magazine/rifle/tx29launcher/he in req since it's nonexistent.

Lumipharon replied:
available from req for now

novaepee reviewed (commented):
Impossible to acquire /obj/item/ammo_magazine/rifle/tx29launcher/he without admin spawn. Either add it in or remove it.

XSlayer300 commented on code/modules/projectiles/guns/rifles.dm:
```diff
@@ -1391,3 +1391,92 @@
+// TX29 rifle
+
+/obj/item/weapon/gun/rifle/tx29
+	name = "\improper TX-29 assault rifle"
```
```suggestion
	name = "\improper TX-29 individual combat weapon unit"
```
gotta make the name fancy tho, but it's up to you.

Lumipharon commented:
Ahhhh, git/vsc shat itself

Lumipharon commented:
I literally took the dmi's from the master, added my sprites to them and merged them again. How is this still conflicting?

Lumipharon commented:
Moved to ops vendor, only 1 crate available round start. Reduced contents to 2 air burst and 2 incend mags. Fixed the req order to actually order the right crate.
Apparently @David-Stormwell is fine with it being here, rather than the standard unga vendor.

TiviPlus commented on code/game/objects/machinery/vending/marine_vending.dm:
```diff
@@ -544,6 +544,8 @@
 	products = list(
 		"Surplus Special Equipment" = list(
 			/obj/item/beacon/supply_beacon = 1,
+			/obj/item/storage/box/tx29_system = 1,
+			/obj/item/ammo_magazine/rifle/tx29 = 6,
```
I reccomend instead of a req implementation to see how the placable MG does it, ie limited amount in pfc weapon vendors

Lumipharon replied:
I had it in the pfc vendor before the last commit, Nova said David asked for it to be in the ops vendor instead though.

Lumipharon replied:
Back in unga vendor. David never actually replied to me when I asked why he wanted it in the ops vendor.

TiviPlus commented on code/game/objects/items/storage/marine_boxes.dm:
```diff
@@ -171,8 +171,24 @@
 	new /obj/item/ammo_magazine/rocket/recoilless/light(src)
 	new /obj/item/storage/backpack/rpg(src)
 
+/obj/item/storage/box/tx29_system
+	name = "\improper TX-29 crate"
+	desc = "A large case containing a TX-29 squad support weapon and its grenade launcher payload. Standard rifle magazines sold separately. Drag this sprite into you to open it up!\nNOTE: You cannot put items back inside this case."
```
Also remove this box, just put all the parts in the vendors seperately

TiviPlus replied:
If people are hoarding ammo it wouldnt matter because the same person would just vend all the boxes and take the ammo, so do it

Lumipharon replied:
Ok it's back in unga vendor without the crate.

TiviPlus reviewed (changes requested):
have you considered davids request to also remove impact grenades in favor of this by the way?

TiviPlus commented:
@Kuro020 review sprites

Lumipharon commented:
> have you considered davids request to also remove impact grenades in favor of this by the way?

Wait, @David-Stormwell you want impact grenades gone entirely?

Kuro020 reviewed (changes requested):
sprite is ok but wtf are these changes
![image](https://user-images.githubusercontent.com/52428261/157551458-f7d95750-23a1-4d46-b990-ae4c25166bb8.png)

Lumipharon commented:
> sprite is ok but wtf are these changes ![image](https://user-images.githubusercontent.com/52428261/157551458-f7d95750-23a1-4d46-b990-ae4c25166bb8.png)

So the grenades only need 1 sprite, not directionals. Projectile code already rotates the projectile sprite to the correct facing, so having direction sprites broke this (basically when you fire a grenade currently, it always faces north)

I uh, didn't touch that blank ammohud icon so dunno what happened there. The only other thing I did was dmi merge conflict fix, if that somehow fucked it?

That anglo_example one I added them because I needed a template to actually see what pixels would be covered when making the inhand/stored gun sprites. I didn't mean to actually add that to the PR, but afaik those example sprites are not actually used in game, so can it stay? Presumably anyone adding/changing gun sprites will have the same issue, and it's handy having the template in file, rather than having to stich body part sprites together.

Lumipharon commented:
Merge conflict went home.

Lumipharon commented:
Removed 25 impact for the 20 airburst nades that are now in roundstart. Gun back in unga vendor.

Lumipharon commented:
Added knockback to the grenade itself. This means direct hits won't always stack ALL the projectiles on the poor fucker. May make it too weak tho.

Al-1ce commented:
The magazines themselves are resprites of the TL-127, - and while I think that's fine, they look strange and don't exactly match the color scheme of any other weapon. This makes them stick out and appear strange for seemingly no reason whatsoever.
![image](https://user-images.githubusercontent.com/71247506/158081630-7fa4932d-9af9-4182-a634-291b021ee979.png)
![image](https://user-images.githubusercontent.com/71247506/158081599-f7eb952c-855d-4616-a13a-d889e52941e7.png)

Take the TX-15 for example. While the magazines are colored separately, they are not neon and easier on the eyes, contributing to the consistency of every other weapon that has colored magazines. While I think the TX-55 mags may be redeemable, if a bit strange looking, the bullets are bizarre and look like they're from an entirely separate codebase.
![image](https://user-images.githubusercontent.com/71247506/158081571-af0bc55a-d82b-436e-9ffb-1c3ea6cc5dca.png)

Shotgun ammunition is the only other type of ammo that has this varying level in color, and that is for easy distinction. If there are only two magazine types for either, or each of these weapons, it isn't consistent or even sensical for them to be bright neon colors. I recommend replacing the current sprites for the bullets with the TL-127 ones, as well as the magazines, with the recolor in a more duller tone. With this, I think they would look overall better. 
![image](https://user-images.githubusercontent.com/71247506/158081729-26851a90-e7c7-46ee-9d8e-233c9810c311.png)
![image](https://user-images.githubusercontent.com/71247506/158081653-f7fe5456-c438-43e6-aa58-102cc4361886.png)
![image](https://user-images.githubusercontent.com/71247506/158081661-c4dadefa-2716-42ad-baee-a8cd56470dd7.png)

Lumipharon commented:
>Shotgun ammunition is the only other type of ammo that has this varying level in color, and that is for easy distinction.

I'm opening to changing them man, but have you looked at the colours of these sort of grenades in real life? They're actually brighter then what I've gone for. I just based them off the colour palette Ocelot used for some rifle grenade sprites he had
![image](https://user-images.githubusercontent.com/7869430/158086831-9d769dc6-e703-4f6e-a2c0-5a2cdb8cef6b.png)

I've also got some other grenade ideas in mind, but no point in bringing them in unless this base PR gets approved.

I'm open to making the grenades/mags less saturated, again I'm not a spriter, I just used colour palettes I had as best as I could (i.e. poorly) so I'm sure someone could make nicer sprites in general, but the size of the nade sprites are fine. They're a 20x28mm grenade, not a bullet.

hyper2snyper commented on code/game/objects/items/storage/marine_boxes.dm:
```diff
@@ -171,11 +171,6 @@
-
-
-
-
```
this is nice but completely out of the way.

Lumipharon replied:
originally TX55 has in a box like RR, but Tivi said scrap the box. You uh, want the spaces back?

hyper2snyper commented on code/game/objects/machinery/vending/marine_vending.dm:
```diff
@@ -108,7 +111,6 @@
 		"Grenades" = list(
 			/obj/item/explosive/grenade = 600,
 			/obj/item/explosive/grenade/m15 = 30,
-			/obj/item/explosive/grenade/impact = 125,
```
why are impacts kill?

Lumipharon replied:
David and Tivi said to do it.

hyper2snyper commented on code/game/objects/machinery/vending/marine_vending.dm:
```diff
@@ -574,7 +576,6 @@
 		),
 		"Grenade Boxes" = list(
 			/obj/item/storage/box/visual/grenade/frag = 2,
-			/obj/item/storage/box/visual/grenade/impact = 2,
```
same here

Lumipharon replied:
As above

hyper2snyper commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -190,6 +190,31 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
+			var/obj/item/weapon/gun/g = source
+			if(source) //Check for the source so we don't runtime if we have bonus projectiles from a non-gun source like a Spitter
+				new_proj.damage *= g.damage_mult //Bonus or reduced damage based on damage modifiers on the gun.
+		else //If no bonus type is defined then the extra projectiles are the same as the main one.
```
early continue

Lumipharon replied:
wat dat

hyper2snyper commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -190,6 +190,31 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
+		var/new_angle = angle + ( (rand(0, main_proj.ammo.bonus_projectiles_scatter)) * (prob(50) ? 1 : -1) )
+		if(new_angle < 0)
+			new_angle += 360
+		else if(new_angle > 360)
```
dont need the else if since new_angle cant both be less than 0 and greater than 360

Lumipharon replied:
making into a simple else.

hyper2snyper commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1143,6 +1168,113 @@ datum/ammo/bullet/revolver/tp44
+
+/datum/ammo/tx54/on_hit_mob(mob/M, obj/projectile/proj)
+	staggerstun(M, proj, stagger = 0, slowdown = 0.5, knockback = 1, shake = 0)
+	var/main_proj_angle = Get_Angle(proj.firer, M)
```
only used once after this, dont need to define a new var here

Lumipharon replied:
What should I do instead?

hyper2snyper commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1143,6 +1168,113 @@ datum/ammo/bullet/revolver/tp44
+	bonus_projectiles_amount = 0
+
+/datum/ammo/tx54/on_hit_obj(obj/O, obj/projectile/proj)
+	var/main_proj_angle = Get_Angle(proj.firer, O)
```
like above

hyper2snyper commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1143,6 +1168,113 @@ datum/ammo/bullet/revolver/tp44
+	bonus_projectiles_amount = 0
+
+/datum/ammo/tx54/on_hit_turf(turf/T, obj/projectile/proj)
+	var/main_proj_angle = Get_Angle(proj.firer, T)
```
again

hyper2snyper commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1143,6 +1168,113 @@ datum/ammo/bullet/revolver/tp44
+	bonus_projectiles_amount = 0
+
+/datum/ammo/tx54/do_at_max_range(obj/projectile/proj)
+	var/main_proj_angle = Get_Angle(proj.firer, get_turf(proj))
```
and here too

hyper2snyper commented on code/modules/projectiles/guns/rifles.dm:
```diff
@@ -1391,3 +1391,117 @@
+//-------------------------------------------------------
+// TX55 rifle
+
+/obj/item/weapon/gun/rifle/tx55
```
you said the rifle was kill tho?

hyper2snyper commented on code/modules/projectiles/projectile.dm:
```diff
@@ -130,6 +130,9 @@
 		if(!(ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE))
 			qdel(src)
 			return
+		if(!(ismob(AM) && CHECK_BITFIELD(ammo.flags_ammo_behavior, AMMO_PASS_THROUGH_MOB)))
```
combine this with the check above

Lumipharon replied:
Making it an ||

hyper2snyper commented on code/modules/projectiles/projectile.dm:
```diff
@@ -444,6 +447,8 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
 				thing_to_uncross.do_projectile_hit(src)
 				if(ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE)
 					continue
+				if(ismob(thing_to_uncross) && CHECK_BITFIELD(ammo.flags_ammo_behavior, AMMO_PASS_THROUGH_MOB))
```
same here

Lumipharon replied:
as above

hyper2snyper commented on code/modules/projectiles/projectile.dm:
```diff
@@ -491,6 +496,8 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
 				thing_to_uncross.do_projectile_hit(src)
 				if(ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE)
 					continue
+				if(ismob(thing_to_uncross) && CHECK_BITFIELD(ammo.flags_ammo_behavior, AMMO_PASS_THROUGH_MOB))
```
again

Lumipharon replied:
as above

hyper2snyper commented on code/modules/projectiles/projectile.dm:
```diff
@@ -514,6 +521,8 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
 				thing_to_uncross.do_projectile_hit(src)
 				if(ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE)
 					continue
+				if(ismob(thing_to_uncross) && CHECK_BITFIELD(ammo.flags_ammo_behavior, AMMO_PASS_THROUGH_MOB))
```
here too

Lumipharon replied:
as above

hyper2snyper commented on code/modules/projectiles/projectile.dm:
```diff
@@ -1062,6 +1075,8 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
 				thing_to_uncross.do_projectile_hit(src)
 				if(ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE)
 					continue
+				if(ismob(thing_to_uncross) && CHECK_BITFIELD(ammo.flags_ammo_behavior, AMMO_PASS_THROUGH_MOB))
```
merge with check above

Lumipharon replied:
as above

hyper2snyper commented on code/modules/projectiles/projectile.dm:
```diff
@@ -1086,6 +1101,8 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
 				thing_to_uncross.do_projectile_hit(src)
 				if(ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE)
 					continue
+				if(ismob(thing_to_uncross) && CHECK_BITFIELD(ammo.flags_ammo_behavior, AMMO_PASS_THROUGH_MOB))
```
again

Lumipharon replied:
as above

hyper2snyper commented on code/modules/projectiles/projectile.dm:
```diff
@@ -1102,6 +1119,8 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
 				thing_to_uncross.do_projectile_hit(src)
 				if(ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE)
 					continue
+				if(ismob(thing_to_uncross) && CHECK_BITFIELD(ammo.flags_ammo_behavior, AMMO_PASS_THROUGH_MOB))
```
another... last one i think?

Lumipharon replied:
aye

Al-1ce commented:
![image](https://user-images.githubusercontent.com/71247506/158914647-38e2b6f1-0761-481e-bc10-895bd9aeb5ba.png)
Description in req is unusually long and out of place.

AGasCanister commented:
> 

it breaks req panel, make it smaller on god

Lumipharon commented:
> it breaks req panel, make it smaller on god

Will do, I meant to change that sorry.

Lumipharon commented:
Last round of changes barring any further feedback/code review.
Added the T49 scope to the TX-54 and TX-55. Not compatible with any other weapon currently.
The scope provides low light visibility when scoped exactly like optical imaging glasses, and enabled aim mode on the gun.
Also acts as a scope.

Significantly, I've coded in the capability for aim fire for attachment weapons where BOTH the main gun and attachment gun have aim mode. This means the TX-55 will be able to aim mode with both firemodes.

Also some slight tweaks to the nades. Added an explosion sound it was lacking, slightly less horrifically expensive at req, and less fall off (currently the AOE damage is pretty soft tbh)

XSlayer300 commented:
Add this weapon to the EORD (with the TX-55) and Vahalla Req if you haven't already.

Lumipharon commented:
> Add this weapon to the EORD (with the TX-55) and Vahalla Req if you haven't already.

Yeah I forgot Valhalla. I'll just do that in a followup PR so I don't have to keep making changes to this one, then ask for TM again.

Lumipharon commented:
Tivi commanded that ammo be made more bountiful, so that xenos shall salt, and ungas shall get skillissued in the back more.
Also changed the scatter back to a fixed pattern to avoid a few edge cases of jankiness.

XSlayer300 commented:
> > Add this weapon to the EORD (with the TX-55) and Vahalla Req if you haven't already.
> 
> Yeah I forgot Valhalla. I'll just do that in a followup PR so I don't have to keep making changes to this one, then ask for TM again.

![image](https://user-images.githubusercontent.com/35091844/160220443-e3c50d89-f65a-4cd8-9bca-1cccef9625fa.png)

(better add another final to it)

Al-1ce commented:
It has a nice niche and fits in well after seeing it in action for a month or two. Cool PR.
